File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        def array_to_tree(left, right):
            if left > right:
                return
            nonlocal preorder_index
            root_val = preorder[preorder_index]
            root = TreeNode(root_val)
            preorder_index += 1
            
            root.left = array_to_tree(left, inorder_map[root_val]-1)
            root.right = array_to_tree(inorder_map[root_val]+1,right)
            
            return root
            
        preorder_index = 0 
        inorder_map = {}
        for idx, val in enumerate(inorder):
            inorder_map[val] = idx
        print(inorder_map)
        left, right = 0, len(inorder)-1
        root = array_to_tree(left, right)
        return root

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int preOrderIdx;
    private Map<Integer, Integer> inorderMap;
    
    public TreeNode helperBuildTree(int left, int right, int[] preorder) {
        if (left > right) return null;

        int rootVal = preorder[preOrderIdx];
        preOrderIdx ++;

        TreeNode root = new TreeNode(rootVal);
        int inorderIdx = inorderMap.get(rootVal);
        root.left = helperBuildTree(left, inorderIdx-1, preorder);
        root.right = helperBuildTree(inorderIdx+1, right, preorder);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        preOrderIdx = 0;
        int left = 0;
        int right = preorder.length - 1;
        inorderMap = new HashMap<>();

        for(int i = 0; i<inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }

        return helperBuildTree(left, right, preorder);
    }
}